flow/layers/physical_model_layer.cc

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/layers/physical_model_layer.h"
+
+#include "third_party/skia/include/utils/SkShadowUtils.h"
+
+namespace flow {
+
+PhysicalModelLayer::PhysicalModelLayer()
+    : rrect_(SkRRect::MakeEmpty()) {}
+
+PhysicalModelLayer::~PhysicalModelLayer() {}
+
+void PhysicalModelLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
+  PrerollChildren(context, matrix);
+  if (!context->child_paint_bounds.intersect(rrect_.getBounds()))
+    context->child_paint_bounds.setEmpty();
+
+  // Add some margin to the paint bounds to leave space for the shadow.
+  SkRect bounds(context->child_paint_bounds);
+  bounds.outset(50.0, 50.0);
+  set_paint_bounds(bounds);
+}
+
+void PhysicalModelLayer::Paint(PaintContext& context) {
+  TRACE_EVENT0("flutter", "PhysicalModelLayer::Paint");
+  FTL_DCHECK(!needs_system_composite());
```

Comment:
It looks like Fuchsia support will require a broader refactoring.  Added a TODO.